core/modules/wcontrol/Foreman.cc

```diff
@@ -95,42 +97,61 @@ Foreman::~Foreman() {
     _pool->shutdownPool();
 }
 
-/// Put the task on the scheduler to be run later.
-void Foreman::processTask(std::shared_ptr<wbase::Task> const& task) {
 
+void Foreman::_setRunFunc(shared_ptr<wbase::Task> const& task) {
     auto func = [this, task](util::CmdData*){
         proto::TaskMsg const& msg = *task->msg;
         int const resultProtocol = 2; // See proto/worker.proto Result protocol
         if (!msg.has_protocol() || msg.protocol() < resultProtocol) {
             LOGS(_log, LOG_LVL_WARN, "processMsg Unsupported wire protocol");
-            if (!task->getCancelled()) {
+            if (!task->isCancelled()) {
                 // We should not send anything back to xrootd if the task has been cancelled.
+                mutex* streamMutex = task->sendChannel->getStreamMutexPtr();
```

Comment:
I'm sorry to say so, but exposing internals mutexes of an object (SendChannel) just to call some method of an object smells like "spaghetti code". This reveals too much details on the internal implementation of `SendChannel` to the outside code and it will for sure make the code more difficult to debug and migrate in the future. Why not to follow a simple design pattern of the "synchronized" methods and acquire the lock inside the method implementation?
```
void SendChannel::sendError(std::string const& msg) {
    std::lock_guard<std::mutex> const lock(_mtx);
    ...
}
```
In case if the external locking is absolutely necessary then my second comment is about the pointer:
```
mutex* streamMutex = ...
```
Wby not to just pass a reference (assuming there is the accessor method returning the reference `std::mutex& sendChannel->getStreamMutex()`):
```
std::lock_guard<std::mutex> streamLock(task->sendChannel->getStreamMutex());
```

Reply:
The locking  of the mutex needs to happen outside of the SendChannel class to use the reliable lock_guard class. I don't think using a ponter over a reference makes much difference. I can try making a function that returns a lock_guard object.